### PR TITLE
Copy paste login

### PR DESCRIPTION
**Pull request by danking**
```
# hailctl auth copy-paste-login LXTFDIXKJ6N4Q23G2Q53VQ7GDOGUTYUUWHQI7EZZ6XHLUKDUFZGQ====                                            
Logged into namespace dking as dking.
```
It is safe to reprint this token because it is not only expired but already used. Tokens live for 5 minutes and can be used only once. You can try logging in with that token (to dking's auth) to see for yourself.

The user page now includes a button to retrieve a fresh copy-paste token:

Screen Shot 2020-04-15 at 11 35 20 PM

This sends you to a rather ugly page that includes a base32 encoded (dashes create word breaks which makes highlighting annoying, so I preferred base32 over base64) token. That token is good for 5 minutes. You copy that token and log in as above.

Screen Shot 2020-04-19 at 4 16 27 PM

### CODE REVIEW

File: auth/auth/auth.py [hunk 1]
```diff
@@ -156,21 +189,18 @@ async def get_users(request, userdata):
...
     post = await request.post()
     username = post['username']
     email = post['email']
     is_developer = post.get('is_developer') == '1'
 
-    async with dbpool.acquire() as conn:
-        async with conn.cursor() as cursor:
-            await cursor.execute(
-                '''
+    user_id = await db.execute_insertone(
```
[the remark this answers is not in the document]
Comment (danking) [reply auth/auth/auth.py:198]: execute_insertone's return statement is: `return cursor.lastrowid`, so this change doesn't change the semantics of the code as it was written.

Moreover, the users table's primary key is the user id, so I think the code as was written was correct.

File: build.yaml [hunk 2]
```diff
@@ -173,6 +173,8 @@ steps:
    migrations:
     - name: initial
       script: /io/sql/initial.sql
+    - name: copy-paste-tokens
+      script: /io/sql/copy-paste-tokens.sql
```
[the remark this answers is not in the document]
Comment (danking) [reply build.yaml:177]: Why do we need to manually restart it? Batch shouldn't need a working auth server to continue to execute an already submitted batch.

File: gear/gear/auth_utils.py [hunk 3]
```diff
@@ -2,25 +2,20 @@
...
 
-async def insert_user(dbpool, spec):
+async def insert_user(db, spec):
     assert all(k in spec for k in ('state', 'username'))
 
-    async with dbpool.acquire() as conn:
-        async with conn.cursor() as cursor:
-            await cursor.execute(
-                f'''
+    return await db.execute_insertone(
```
[the remark this answers is not in the document]
Comment (danking) [reply gear/gear/auth_utils.py:8]: see above, I'm just exactly replicating the existing code using the `Database` interface instead.

File: hail/python/hailtop/auth/auth.py [hunk 4]
```diff
@@ -35,3 +36,36 @@ def namespace_auth_headers(deploy_config, ns, authorize_target=True):
...
+        async with session.post(deploy_config.url('auth', '/api/v1alpha/copy-paste-login'),
+                                params={'copy_paste_token': copy_paste_token}) as resp:
+            resp = await resp.json()
+    token = resp['token']
+    username = resp['username']
+
+    tokens = get_tokens()
+    tokens[auth_ns] = token
+    dot_hail_dir = os.path.expanduser('~/.hail')
+    if not os.path.exists(dot_hail_dir):
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hailtop/auth/auth.py:67]: This is used by `hailctl auth login`. The two options are:
    <p>
       If you need to authenticate a Jupyter Notebook session (for example, a
       Terra Jupyter Notebook), copy and paste this into your notebook:
     </p>
     <pre>from hailtop.auth import copy_paste_login; copy_paste_login('{{ copy_paste_token }}')</pre>
     <p>
       If you need to authenticate from a terminal, copy and paste this into your
       terminal:
     </p>
     <pre>hailctl auth copy-paste-login "{{ copy_paste_token }}"</pre>

**Comment by danking** [comment]
The auth service is used extensively, as you mention. I also added tests for my new functionality. I doubt the PR would pass if auth wasn't correct.

File: build.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jigold) [reply build.yaml:177]: oh, you're right. The next deploy step after the database will redeploy it.

**Comment by jigold** [comment]
This looks good, but requires a rebase and the base52 addition I believe.

**Comment by danking** [comment]
addressed

**Comment by jigold** [comment]
This needs a rebase

File: hail/python/setup-hailtop.py [hunk 5]
```diff
@@ -1,11 +1,30 @@
...
-    version = '0.0.1',
-    url = 'https://github.com/hail-is/hail.git',
-    author = 'Hail Team',
-    author_email = 'hail@broadinstitute.org',
-    description = 'Toplevel Hail module',
-    packages = find_packages()
+    name='hailtop',
+    version="0.0.1",
+    author="Hail Team",
+    author_email="hail-team@broadinstitute.org",
```
Comment (jigold) [inline hail/python/setup-hailtop.py:9]: This should be hail@broad...

Comment (danking) [reply hail/python/setup-hailtop.py]: The `hail` PyPI package uses `hail-team@`. I changed them both to `hail@`.

**Comment by jigold** [comment]
Tests are failing.

**Pull request merged**